### ISSUE TITLE
CUDA files compile error inside native folder

### ISSUE THREAD

**Issue by timemeansalot** [matter-labs/era-boojum-cuda#34]
### 🐛 Bug Report
I cloned this repo, I can not compile the cuda files inside the native filder.

#### 📝 Description

I cloned the repo, then I use `cargo build` to compile all the files, when it compiles the `native/ntt_b2n.cuh` file, the compiler says `/zksync/era-boojum-cuda/native/ntt_b2n.cuh(345): error: initializer not allowed for __shared__ variable`. I want to know if [this code line](https://github.com/matter-labs/era-boojum-cuda/blob/1e652c57c54c5de53fb450ee882bedfeeecf63ac/native/ntt_b2n.cuh#L14) has bug, or I used the wrong way to compile them.

#### 🔄 Reproduction Steps

I tried the following 2 ways to compile the code, they all told the same error `/zksync/era-boojum-cuda/native/ntt_b2n.cuh(345): error: initializer not allowed for __shared__ variable`.

1. The rust way: I cloned this repo, and I simple use cargo to build the project by `cargo build`
2. The cmake way: I cloned this repo, then I go to the native folder, I create a folder named build, enter the `build` folder and use `cmake .. && make` command to compile all the cuda files.

As I said, both the 2 method reported the same error `/native/ntt_b2n.cuh(14): error: initializer not allowed for __shared__ variable`

By the way, I am using nvcc version 12.3, g++ version 11.4.0

#### 🤔 Expected Behavior

I hope I can build this project successfully by simple using `cargo build`, I am not sure  this is the cuda codes bug or because I use the wrong way to compile them. 
#### 😯 Current Behavior

Describe what actually happened.

#### 🖥️ Environment
1. nvcc --version
  nvcc: NVIDIA (R) Cuda compiler driver
  Copyright (c) 2005-2023 NVIDIA Corporation
  Built on Fri_Sep__8_19:17:24_PDT_2023
  Cuda compilation tools, release 12.3, V12.3.52
  Build cuda_12.3.r12.3/compiler.33281558_0
2. g++ --version
g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.

#### 📋 Additional Context

Add any other context about the problem here. If applicable, add screenshots to help explain.

#### 📎 Log Output
I used the following command to compile the codes
```
git clone git@github.com:matter-labs/era-boojum-cuda.git
cd era-boojum-cuda/native/
mkdir build && cd build
cmake ..
make
```

And the main error message I got is:

```
[  5%] Building CUDA object CMakeFiles/boojum-cuda-native.dir/ntt.cu.o
/home/jie/Developer/zksync/era-boojum-cuda/native/ntt_b2n.cuh(14): error: initializer not allowed for __shared__ variable

/home/jie/Developer/zksync/era-boojum-cuda/native/ntt_b2n.cuh(14): error: initializer not allowed for __shared__ variable

/home/jie/Developer/zksync/era-boojum-cuda/native/ntt_b2n.cuh(128): error: initializer not allowed for __shared__ variable

/home/jie/Developer/zksync/era-boojum-cuda/native/ntt_b2n.cuh(345): error: initializer not allowed for __shared__ variable

/home/jie/Developer/zksync/era-boojum-cuda/native/ntt_n2b.cuh(15): error: initializer not allowed for __shared__ variable

/home/jie/Developer/zksync/era-boojum-cuda/native/ntt_n2b.cuh(15): error: initializer not allowed for __shared__ variable

/home/jie/Developer/zksync/era-boojum-cuda/native/ntt_n2b.cuh(134): error: initializer not allowed for __shared__ variable

/home/jie/Developer/zksync/era-boojum-cuda/native/ntt_n2b.cuh(339): error: initializer not allowed for __shared__ variable
```

**Comment by mcarilli** [comment]
That's strange. The two lines you mentioned ([14](https://github.com/matter-labs/era-boojum-cuda/blob/1e652c57c54c5de53fb450ee882bedfeeecf63ac/native/ntt_b2n.cuh#L14), [345](https://github.com/matter-labs/era-boojum-cuda/blob/1e652c57c54c5de53fb450ee882bedfeeecf63ac/native/ntt_b2n.cuh#L345)) don't contain initializers, they just allocate empty shared memory, so I think it's an nvcc bug.

I also can't reproduce the failure locally. I can build main in my local environment, which is similar to yours except for a later release of CUDA 12.3:
```
mcarilli:era-boojum-cuda$ git branch -v
* main            1e652c5 bump rust-toolchain and dependencies (#32)
```
```
mcarilli:~$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Fri_Nov__3_17:16:49_PDT_2023
Cuda compilation tools, release 12.3, V12.3.103
Build cuda_12.3.r12.3/compiler.33492891_0
```
```
mcarilli:~$ g++ --version
g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Since the issue appears to be an nvcc bug fixed by newer versions, I recommend you update to the latest CUDA toolkit release and try again.

If you're not able to update cuda, let me know and we can try to find another workaround.

**Comment by timemeansalot** [comment]
Thanks a lot for the reply. 

This bug is turelly about the nvcc compiler, the more detaild information is:

1. I have multiple cuda on my computer, the oldest one is cuda version 11. I set cuda version to 12.3 in my PATH variable, so the terminal will output that I am using nvcc 12.3.
2. the problem is that CMAKE can not choose the correct cuda which I set in my PATH varaible. In the CMakeLists.txt file inside `native` build, I have to add `set(CMAKE_CUDA_COMPILER /usr/local/cuda-12.3/bin/nvcc)` before `project(boojum-cuda-native)
` so that CMAKE can find the correct cuda on my system.

Thanks a again for your reply :). I think I can close this issue now~
